trieste/models/keras/models.py

```diff
@@ -0,0 +1,303 @@
+            ``query_points``.
+        """
+        query_points_transformed = self.prepare_data(query_points)
+
+        ensemble_distributions = self.model(query_points_transformed)
+        predicted_means = tf.math.reduce_mean(
+            [dist.mean() for dist in ensemble_distributions], axis=0
+        )
+        predicted_vars = (
+            tf.math.reduce_mean(
+                [dist.variance() + dist.mean() ** 2 for dist in ensemble_distributions], axis=0
+            )
+            - predicted_means ** 2
+        )
+
+        return predicted_means, predicted_vars
+
+    def predict_ensemble(self, query_points: TensorType) -> list[tuple[TensorType, TensorType]]:
+        """
+        Returns mean and variance at ``query_points`` for each member of the ensemble.
+
+        This method assumes that the final layer in each member of the ensemble is
+        probabilistic, an instance of :class:`¬tfp.distributions.Distribution`, in particular
+        `mean` and `variance` methods should be available.
+
+        :param query_points: The points at which to make predictions.
+        :return: The predicted mean and variance of the observations at the specified
+            ``query_points`` for each member of the ensemble.
+        """
+        query_points_transformed = self.prepare_data(query_points)
+
+        ensemble_distributions = self.model(query_points_transformed)
+        predicted_means_vars = [(dist.mean(), dist.variance()) for dist in ensemble_distributions]
+
+        return predicted_means_vars
+
+    def sample(self, query_points: TensorType, num_samples: int) -> TensorType:
+        """
+        Return ``num_samples`` samples at ``query_points``.
+
+        We use the mixture approximation in :meth:`predict` for ``query_points`` and
+        sample ``num_samples`` times from a Gaussian distribution given by the predicted
+        mean and variance.
+
+        :param query_points: The points at which to sample, with shape [..., N, D].
+        :param num_samples: The number of samples at each point.
+        :return: The samples. For a predictive distribution with event shape E, this has shape
+            [..., S, N] + E, where S is the number of samples.
+        """
+
+        predicted_means, predicted_vars = self.predict(query_points)
+        normal = tfp.distributions.Normal(predicted_means, tf.sqrt(predicted_vars))
+        samples = normal.sample(num_samples)
+
+        return samples  # [num_samples, len(query_points), 1]
+
+    # def trajectory_sampler(self) -> TrajectorySampler:
+    #     """
+    #     Return a trajectory sampler. For :class:`DeepEnsemble`, we use an ensemble
+    #     sampler that randomly picks a network from the ensemble and uses it for generating a
+    #     trajectory.
+
+    #     :return: The trajectory sampler.
+    #     """
+    #     return EnsembleTrajectorySampler(self.model, self.ensemble_size)
```

Comment:
eh?

Reply:
ah, forgot to delete that part, should be part of the next PR